### PR TITLE
Fix url shortening

### PR DESCRIPTION
**Pull request by cco3**
This change also includes a big refactoring commit that turns the MetadataResolver into a generic PwsClient.  The second commit offloads URL shortening to that new generic client.

### CODE REVIEW

**Comment by cco3** [comment]
Paging @g-ortuno if you have time to take a look

**Comment by mmocny** [comment]
I can take a look in a few hours as well.  Not an expert in our Android app, but should be okay here..

**Comment by mmocny** [comment]
Confirmed this is working well, woot!

File: android/PhysicalWeb/app/src/main/AndroidManifest.xml [hunk 1]
```diff
@@ -16,6 +16,7 @@
 
     <application
         android:allowBackup="true"
+        android:name=".PhysicalWeb"
```
Comment (mmocny) [inline android/PhysicalWeb/app/src/main/AndroidManifest.xml]: Is the `.` intentional?

File: android/PhysicalWeb/app/src/main/java/org/physical_web/physicalweb/PwsClient.java [hunk 2]
```diff
@@ -37,29 +37,33 @@
...
+ * For the largest use, metadata requests, the physical web service
+ * scrapes the page at the given url for its metadata.
  */
-
-class MetadataResolver {
-  private static final String TAG = "MetadataResolver";
-  private static final String METADATA_URL = "http://url-caster.appspot.com/resolve-scan";
-  //private static final String METADATA_URL = "http://url-caster-dev.appspot.com/resolve-scan";
-  private static final String DEMO_METADATA_URL = "http://url-caster.appspot.com/demo";
+class PwsClient {
```
Comment (mmocny) [inline android/PhysicalWeb/app/src/main/java/org/physical_web/physicalweb/PwsClient.java]: May we scan the codebase for references to this old class.. I think it's used in some docs.

File: android/PhysicalWeb/app/src/main/java/org/physical_web/physicalweb/PwsClient.java [hunk 3]
```diff
@@ -37,29 +37,33 @@
...
+  private static final String TAG = "PwsClient";
+  private static final String PROD_URL = "http://url-caster.appspot.com";
+  //private static final String DEV_URL = "http://url-caster-dev.appspot.com";
```
Comment (mmocny) [inline android/PhysicalWeb/app/src/main/java/org/physical_web/physicalweb/PwsClient.java]: Drive-by future feature request: iOS switches to DEV backend when in dev mode.. can we implement something like this for Android, so you can test PWS features without an App re-build?

File: android/PhysicalWeb/app/src/main/java/org/physical_web/physicalweb/PwsClient.java [hunk 4]
```diff
@@ -37,29 +37,33 @@
...
+  private static final String RESOLVE_SCAN_PATH = "resolve-scan";
+  private static final String DEMO_RESOLVE_SCAN_PATH = "demo";
+  private static final String SHORTEN_URL_PATH = "shorten-url";
   private static final int UNDEFINED_SCORE = -1;
```
Comment (mmocny) [inline android/PhysicalWeb/app/src/main/java/org/physical_web/physicalweb/PwsClient.java]: Drive-by bug report: We don't use score, we use "rank".

File: android/PhysicalWeb/app/src/main/AndroidManifest.xml (hunk 1 above)
Comment (cco3) [inline android/PhysicalWeb/app/src/main/AndroidManifest.xml]: Yea, see .MainActivity on L24

File: android/PhysicalWeb/app/src/main/java/org/physical_web/physicalweb/PwsClient.java (hunk 3 above)
Comment (cco3) [inline android/PhysicalWeb/app/src/main/java/org/physical_web/physicalweb/PwsClient.java]: dev mode?  You mean like when you have developer options enabled on android?  I have no idea if you can even see that, but I'd be willing to look into it.

File: android/PhysicalWeb/app/src/main/java/org/physical_web/physicalweb/PwsClient.java (hunk 2 above)
Comment (cco3) [inline android/PhysicalWeb/app/src/main/java/org/physical_web/physicalweb/PwsClient.java]: Done

**Comment by g-ortuno** [comment]
There is a bug similar to the bug you had last time. If you get the shortened url after the activity gets destroyed the app crashes. You can test it with a slow network:

1. Open config mode
2. Enter long url
3. Save long url
4. Press back before the URL gets shortened.
5. App crashes

Crash log:
```java
java.lang.NullPointerException: Attempt to invoke virtual method 'void org.uribeacon.config.GattRequestQueue.add(android.bluetooth.BluetoothGatt, org.uribeacon.config.GattRequestQueue$RequestType, android.bluetooth.BluetoothGattCharacteristic)' on a null object reference
            at org.uribeacon.config.GattService.writeCharacteristic(GattService.java:85)
            at org.uribeacon.config.ProtocolV1.writeUriBeacon(ProtocolV1.java:92)
            at org.uribeacon.config.UriBeaconConfig.writeUriBeacon(UriBeaconConfig.java:97)
            at org.physical_web.physicalweb.BeaconConfigFragment.setUriBeaconUrl(BeaconConfigFragment.java:347)
            at org.physical_web.physicalweb.BeaconConfigFragment.access$900(BeaconConfigFragment.java:67)
            at org.physical_web.physicalweb.BeaconConfigFragment$5.onUrlShortened(BeaconConfigFragment.java:372)
            at org.physical_web.physicalweb.PwsClient$1.onResponse(PwsClient.java:119)
            at org.physical_web.physicalweb.PwsClient$1.onResponse(PwsClient.java:108)
            at com.android.volley.toolbox.JsonRequest.deliverResponse(JsonRequest.java:65)
            at com.android.volley.ExecutorDelivery$ResponseDeliveryRunnable.run(ExecutorDelivery.java:99)
```

File: android/PhysicalWeb/app/src/main/java/org/physical_web/physicalweb/PwsClient.java (hunk 3 above)
Comment (mmocny) [inline android/PhysicalWeb/app/src/main/java/org/physical_web/physicalweb/PwsClient.java]: I meant an application-level dev mode.  I haven't seen it, but I think there is an option from the dropdown on iOS to enter dev/debug mode or something

Comment (g-ortuno) [inline android/PhysicalWeb/app/src/main/java/org/physical_web/physicalweb/PwsClient.java]: I know @matthewsibigtroth added a dev mode as well. If you long press the url list in android it goes into dev mode.

Comment (cco3) [inline android/PhysicalWeb/app/src/main/java/org/physical_web/physicalweb/PwsClient.java]: Oh, OK.

Comment (mmocny) [inline android/PhysicalWeb/app/src/main/java/org/physical_web/physicalweb/PwsClient.java]: @g-ortuno Oh, sweet!

**Comment by cco3** [comment]
Fixed using volley's tag-based cancellation.

**Comment by g-ortuno** [comment]
Any reason why are you using this instead of the singleton pattern? It seems the singleton pattern is the recommended way:

https://developer.android.com/training/volley/requestqueue.html#singleton

**Comment by cco3** [comment]
Yes, I made PwsClient a singleton.

**Comment by cco3** [comment]
But if the requestqueue is in a singleton, that means it can't know what activities it gets requests from...that's why you use tag-based cancellation.

**Comment by g-ortuno** [comment]
I mean why are you subclassing Application, which seems to be discouraged, and not using a static singleton class?

https://developer.android.com/training/volley/requestqueue.html#singleton

**Comment by cco3** [comment]
I don't create the RequestQueue in the subclassed Application.  I'm assuming that's discouraged because you shouldn't create it until it's needed.

**Comment by cco3** [comment]
Oh, I see in the link to the Application documentation that they prefer you pass along context...

OK, I'll do it that way, but that seems like a pain.

**Comment by g-ortuno** [comment]
I don't really have a preference. I was just curious why you picked one over the other. If you think that the current pattern is better suited for this then LGTM

**Comment by cco3** [comment]
OK, fixed it up.  Tested, everything works.

**Comment by g-ortuno** [comment]
Feel free to merge once travis finishes.
